stan/math/prim/err/elementwise_check.hpp

```diff
@@ -0,0 +1,136 @@
+#ifndef STAN_MATH_PRIM_ERR_ELEMENTWISE_ERROR_CHECKER_HPP
+#define STAN_MATH_PRIM_ERR_ELEMENTWISE_ERROR_CHECKER_HPP
+
+#include <stan/math/prim/err/throw_domain_error.hpp>
+#include <stan/math/prim/fun/get.hpp>
+#include <stan/math/prim/fun/size.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <string>
+#include <sstream>
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+template <typename T_is_good, typename T_exception>
+struct Checker {
+  const T_is_good& is_good;
+  const char* function;
+  const char* name;
+  const char* suffix;
+
+  void raise_error_impl(std::stringstream& ss) { throw T_exception{ss.str()}; }
+
+  template <typename... M>
+  void raise_error_impl(std::stringstream& ss, const char* message,
+                        const M&... messages) {
+    ss << message;
+    raise_error_impl(ss, messages...);
+  }
+
+  template <typename... M>
+  void raise_error_impl(std::stringstream& ss, double value_of_x,
+                        const M&... messages) {
+    ss << value_of_x;
+    raise_error_impl(ss, messages...);
+  }
+
+  template <typename... M>
+  void raise_error_impl(std::stringstream& ss, size_t index,
+                        const M&... messages) {
+    ss << index + 1;
+    raise_error_impl(ss, messages...);
+  }
+
+  template <typename... M>
+  void raise_error(const M&... messages) {
+    std::stringstream ss{};
+    raise_error_impl(ss, messages...);
+  }
+
+  template <typename T, typename = require_stan_scalar_t<T>,
+            typename... T_indices>
+  void check(const T& x, T_indices... indices) {
+    double xd = value_of_rec(x);
+    if (!is_good(xd))
+      raise_error(function, ": ", name, indices..., " is ", xd, suffix);
+  }
+
+  template <typename T, typename = require_eigen_vector_t<T>, typename = void,
+            typename... T_indices>
+  void check(const T& x, T_indices... indices) {
+    for (size_t i = 0; i < stan::math::size(x); ++i)
+      check(x(i), indices..., "[", i, "]");
+  }
+
+  template <typename T, typename... T_indices>
+  void check(const std::vector<T>& x, T_indices... indices) {
+    for (size_t i = 0; i < stan::math::size(x); ++i)
+      check(x[i], indices..., "[", i, "]");
+  }
+
+  template <typename T, typename... T_indices>
+  void check(const Eigen::DenseBase<T>& x, T_indices... indices) {
+    for (size_t n = 0; n < x.cols(); ++n)
+      for (size_t m = 0; m < x.rows(); ++m)
+        check(x(m, n), indices..., "[row=", m, ", col=", n, "]");
+  }
+};
+
+template <typename T_is_good>
+struct Iser {
+  const T_is_good& is_good;
+
+  template <typename T, typename = require_stan_scalar_t<T>>
+  bool is(const T& x) {
+    return is_good(value_of_rec(x));
+  }
+
+  template <typename T, typename = require_not_stan_scalar_t<T>,
+            typename = void>
```

Comment:
[question]
Is this dummy just so `require` and `require_not` can be used with the same template signature?

Reply:
Yes.